### PR TITLE
Queries with inline data (p2p only)

### PR DESCRIPTION
**Pull request by yusefnapora**
This adds a `remoteQueryWithData` method to the `MediachainNode` class, which will fetch multihash `object` references in query results from the remote peer and replace the multihash with `{key: multihashValue, data: actualObject}`

<details>
<summary>An example at the repl</summary>

```
א > remote.queryWithData('SELECT * FROM scratch.* LIMIT 1')
[ { simple: { stmt: [Object] } } ]
א > var res = _[0].simple.stmt
א > console.dir(res, {depth: 100})
{ id: '4XTTM6XcpMB8N8Tnkhc66DqvAotgKzRJgrZbF7SBByihtPnV8:1478207997:34006',
  publisher: '4XTTM6XcpMB8N8Tnkhc66DqvAotgKzRJgrZbF7SBByihtPnV8',
  namespace: 'scratch.moma',
  body: 
   { simple: 
      { object: 
         { key: 'QmPEeMwnRhJB1YV4sZsJV3wNtHz8jsEw1WgzjC8quqxfqU',
           data: 
            { schema: { '/': 'QmXciNhn4P6veuojsR6uRGHBu27pAztp5o1rtMkkmsTtUf' },
              data: 
               { ArtistBio: 'American, 1930–1992',
                 BeginDate: 1930,
                 ConstituentID: 1,
                 DisplayName: 'Robert Arneson',
                 EndDate: 1992,
                 Gender: 'Male',
                 Nationality: 'American',
                 ULAN: null,
                 'Wiki QID': null } } },
        refs: [ 'moma:artist:1' ],
        tags: [],
        deps: [ 'QmXciNhn4P6veuojsR6uRGHBu27pAztp5o1rtMkkmsTtUf' ] } },
  timestamp: 1478207997,
  signature: 
   Buffer [ ... ] }
```

</details>

TODO: 

I'd like to add a new subcommand to the aleph cli that executes a query string against a remote peer and pretty prints the results, so we can get a nice one-liner to use in the Attribution Engine.

### CODE REVIEW

**Comment by parkan** [comment]
could add an invocation like `aleph exec "remote.queryWithData('SELECT * FROM scratch.* LIMIT 1')"`?

**Comment by yusefnapora** [comment]
That could work, although the repl's default printing of the query results isn't to great.  For the example above you'd just get `[ { simple: { stmt: [Object] } } ]`

That's something we should probably tweak regardless though, to make working at the repl more useful.

File: src/peer/repl/index.js [hunk 1]
```diff
@@ -15,16 +28,18 @@ require('yargs')
       // this is the only way to get useful config in subcommands
       'config': true,
       'global': true,
-      'default': home + 'config.json'
+      'default': path.join(home, 'config.json')
```
Comment (parkan) [inline src/peer/repl/index.js]: good call

File: src/peer/repl/index.js [hunk 2]
```diff
@@ -15,16 +28,18 @@ require('yargs')
...
-      'default': home + 'identity.node'
+      'default': path.join(home, 'identity.node')
     },
     'remotePeer': {
       'global': true,
       'type': 'string',
       'describe': 'The remote peer to pair with (multiaddress)'
     }
   })
+  .coerce('config', filePath => ensureDirExists(filePath, 'config file'))
```
Comment (parkan) [inline src/peer/repl/index.js]: you don't want to just put this into the `config` key above as `coerce : { filepath: ... `?

Comment (yusefnapora) [inline src/peer/repl/index.js]: oh sure, that does sound better.

File: src/peer/repl/util.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+function binaryToB64 (result: Object): Object {
+  const replacer = obj => {
+    if (obj instanceof Buffer) {
+      return obj.toString('base64')
+    }
+  }
+
+  return _.cloneDeepWith(result, replacer)
```
Comment (parkan) [inline src/peer/repl/util.js]: all the lodash pieces are broken out so you can just import `lodash.clonedeepwith` but it's not critical https://www.npmjs.com/package/lodash.clonedeepwith

File: src/peer/util.js [hunk 4]
```diff
@@ -177,6 +179,66 @@ function promiseTimeout<T> (timeout: number, promise: Promise<T>): Promise<T> {
   })])
 }
 
+function flatMap<T, U> (array: Array<T>, f: (x: T) => Array<U>): Array<U> {
```
Comment (parkan) [inline src/peer/util.js]: if only Thenables were monadic 😞

File: src/peer/util.js [hunk 5]
```diff
@@ -177,6 +179,66 @@ function promiseTimeout<T> (timeout: number, promise: Promise<T>): Promise<T> {
...
+}
+
+function expandQueryResult (result: QueryResultValueMsg, dataObjects: Array<DataObjectMsg>): Object {
+  const objectMap = {}
+  for (const obj of dataObjects) {
+    let val: Object | string
+    try {
+      val = decode(obj.data)
+    } catch (err) {
+      val = obj.data.toString('base64')
```
Comment (parkan) [inline src/peer/util.js]: might be nice to add a comment here explaining the logic

File: src/peer/util.js [hunk 6]
```diff
@@ -177,6 +179,66 @@ function promiseTimeout<T> (timeout: number, promise: Promise<T>): Promise<T> {
...
+    }
+    objectMap[obj.key] = val
+  }
+
+  function replacer (value: any) {
+    const key = _.get(value, 'object')
+    const data = objectMap[key]
+    if (data != null) {
+      return _.set(value, 'object', {key, data})
```
Comment (parkan) [inline src/peer/util.js]: ah I guess we have all these other lodash calls too, let's take the whole package then

File: src/peer/util.js (hunk 5 above)
Comment (yusefnapora) [inline src/peer/util.js]: hmm, actually I might remove the base64 conversion here.  After I wrote this, I added a pass to convert all `Buffer`s to base64 just before printing the results.  That's a better place for it than here, I think

**Pull request merged**